### PR TITLE
Feature/1533/add color picker to change color

### PR DESCRIPTION
**Pull request by shaman-apprentice**
# Make color of buildings and edges customizable

Closes: #1533

## Description

- Quite a lot changes already and should be sufficient
  - As cleanup I created #1648

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/)
 
 ### Added 🚀
 
+-   Allow color of buildings to be customizable in the ribbon bar and in the legend. Colors of edges are now customizable as well ([#1533](https://github.com/MaibornWolff/codecharta/issues/1533))
```
Comment (ce-bo) [inline CHANGELOG.md:12]: What about the "selected" color and the folder marking colors from the context menu? :)
Does a user know what we mean with "ribbon bar"? Maybe we can add an self-explaining screenshot for that?

File: visualization/app/codeCharta/ui/codeMap/threeViewer/threeSceneService.ts [hunk 2]
```diff
@@ -58,6 +60,11 @@ export class ThreeSceneService implements CodeMapPreRenderServiceSubscriber {
 		this.scene.add(this.lights)
 	}
 
+	onMapColorsChanged(mapColors: MapColors) {
+		this.folderLabelColorSelected = mapColors.selected
+		this.numberOrangeColor = ColorConverter.convertHexToNumber(this.folderLabelColorSelected)
```
Comment (ce-bo) [inline visualization/app/codeCharta/ui/codeMap/threeViewer/threeSceneService.ts:65]: Is it still orange => `this.numberOrangeColor`?

**Comment by shaman-apprentice** [comment]
`appSettings.invertColorRange`, `appSettings.whiteColorBuildings` and all related actions, reducer and subscriber can be probably removed after this PR. I propose to check next year one more time and create a clean up issue, if this is really the case

File: CHANGELOG.md (hunk 1 above)
Comment (shaman-apprentice) [reply CHANGELOG.md:12]: image
upload here, for referring it in changelog 😇

File: visualization/app/codeCharta/ui/codeMap/threeViewer/threeSceneService.ts (hunk 2 above)
Comment (shaman-apprentice) [reply visualization/app/codeCharta/ui/codeMap/threeViewer/threeSceneService.ts]: I renamed all occurrences

File: CHANGELOG.md (hunk 1 above)
Comment (shaman-apprentice) [reply CHANGELOG.md]: The screenshot is a great idea and I added the above one 👍 

- For me "color of buildings" includes "selected" and should be more clear with the screenshot now - or do you have a better suggestion @ce-bo ? :)
- @ce-bo Do you mean the following screenshot with _"the folder marking colors from the context menu"_? I don't remember, if we wanted to update this color picking in this FR or in a new one - guess it makes sense to also do it in this PR (@BridgeAR?)
image

**Comment by Gaugi28** [comment]
>   * On hover no brush is shown, but only on selection brush is added. @Gaugi28 this should be fine, isn't it? 
>     image
> 
@shaman-apprentice The brush should appear if the user hovers over the color icon so she or he can see the possibility for editing the color. 
If the color icon is dark you can't see the brush well. Could the brush be white, if the color icon is dark, and be black, if the color icon is bright?

https://webaim.org/resources/contrastchecker/ Here you can check the contrast. You can find more information here: https://webaim.org/articles/contrast/#ratio

**Comment by shaman-apprentice** [comment]
> @shaman-apprentice The brush should appear if the user hovers over the color icon so she or he can see the possibility for editing the color.
> If the color icon is dark you can't see the brush well. Could the brush be white, if the color icon is dark, and be black, if the color icon is bright?

@Gaugi28 very good point 👍 Do you have a threshold in mind, when it should be black and when white? (ideal related to rgb color - I am not an expert regarding colors...).

I wonder if it is better to invert the color of the brush related to the color picker?

File: visualization/app/codeCharta/util/algorithm/streetLayout/horizontalStreet.ts [hunk 3]
```diff
@@ -174,7 +174,7 @@ export default class HorizontalStreet extends Street {
 				sum += child.width
 			} else {
 				if (child instanceof VerticalStreet) {
-					(<VerticalStreet>child).orientation = VerticalOrientation.DOWN
+					;(<VerticalStreet>child).orientation = VerticalOrientation.DOWN
```
Comment (Med-H) [inline visualization/app/codeCharta/util/algorithm/streetLayout/horizontalStreet.ts:177]: I guess there is no need for the ;

File: visualization/app/codeCharta/util/algorithm/streetLayout/verticalStreet.ts [hunk 4]
```diff
@@ -128,7 +128,7 @@ export default class VerticalStreet extends Street {
 		for (const child of children) {
 			if (sum < totalLength / 2) {
 				if (child instanceof HorizontalStreet) {
-					(<HorizontalStreet>child).orientation = HorizontalOrientation.LEFT
+					;(<HorizontalStreet>child).orientation = HorizontalOrientation.LEFT
```
Comment (Med-H) [inline visualization/app/codeCharta/util/algorithm/streetLayout/verticalStreet.ts:131]: Same as above :)

**Review by Med-H — COMMENTED** [review]
It looks good :)

File: visualization/app/codeCharta/util/algorithm/streetLayout/horizontalStreet.ts (hunk 3 above)
Comment (shaman-apprentice) [reply visualization/app/codeCharta/util/algorithm/streetLayout/horizontalStreet.ts:177]: I agree. In general I find a `;` at the beginning of a line irritating. It is added by our auto formatting 💁‍♂️

**Comment by Gaugi28** [comment]
> > @shaman-apprentice The brush should appear if the user hovers over the color icon so she or he can see the possibility for editing the color.
> > If the color icon is dark you can't see the brush well. Could the brush be white, if the color icon is dark, and be black, if the color icon is bright?
> 
> @Gaugi28 very good point 👍 Do you have a threshold in mind, when it should be black and when white? (ideal related to rgb color - I am not an expert regarding colors...).
> 
> I wonder if it is better to invert the color of the brush related to the color picker?

Here you can find a solution for dynamically changing the color based on background color: https://gomakethings.com/dynamically-changing-the-text-color-based-on-background-color-contrast-with-vanilla-js/

**Comment by BridgeAR** [comment]
Here is the algorithm also described: https://www.w3.org/TR/WCAG20-TECHS/G18.html#G18-tests

**Comment by BridgeAR** [comment]
@shaman-apprentice is this still a draft?

**Comment by shaman-apprentice** [comment]
> @shaman-apprentice is this still a draft?

I changed it to draft again, as I am figuring out a nice way for the brush handling. Should be out of draft again before morning.

**Review by Med-H — APPROVED** [review]
Should be good :)

**Pull request merged**